libraries/chain/controller.cpp

```diff
@@ -354,14 +355,55 @@ struct assembled_block {
                                    },
                                    [&](assembled_block_if& ab) {
                                       auto bsp = std::make_shared<block_state>(ab.bhs, std::move(ab.trx_metas),
-                                                                               std::move(ab.trx_receipts), ab.qc, signer,
+                                                                               std::move(ab.trx_receipts), ab.valid, ab.qc, signer,
                                                                                valid_block_signing_authority);
                                       return completed_block{std::move(bsp)};
                                    }},
                         v);
    }
 };
 
+// A utility to build a valid structure from the parent block
+static valid_t build_valid_structure(const block_state_ptr parent_bsp, const block_header_state& bhs) {
+   assert(parent_bsp);
+   assert(bhs.core.last_final_block_num() >= parent_bsp->core.last_final_block_num());
+   assert(parent_bsp->valid);
+   assert(parent_bsp->valid->finality_mroots.size() == (parent_bsp->block_num() - parent_bsp->core.last_final_block_num() + 1));
+
+   valid_t valid;
+
+   // Copy parent's finality_merkel_tree and finality_mroots.
+   // For finality_mroots, removing any roots from the front end
+   // to block whose block number is last_final_block_num - 1
+   auto start = bhs.core.last_final_block_num() - parent_bsp->core.last_final_block_num();
+   valid = valid_t {
+      .finality_merkel_tree = parent_bsp->valid->finality_merkel_tree,
+      .finality_mroots = { parent_bsp->valid->finality_mroots.cbegin() + start,
+      parent_bsp->valid->finality_mroots.cend() }
+   };
+
+   // append the root of the parent's Validation Tree.
+   valid.finality_mroots.emplace_back(parent_bsp->valid->finality_merkel_tree.get_root());
+
+   // post condition of finality_mroots
+   assert(valid.finality_mroots.size() == (bhs.block_num() - bhs.core.last_final_block_num() + 1));
+
+   // construct block's finality leaf node.
+   valid_t::finality_leaf_node_t leaf_node{
+      .block_num       = bhs.block_num(),
+      .finality_digest = bhs.compute_finalizer_digest(),
+      .action_mroot    = bhs.finality_mroot()
```

Comment:
We also need to make sure the action_mroot here contains the merkle root of the action receipt digests tree.

This is calculated here :

https://github.com/AntelopeIO/leap/blob/main/libraries/chain/controller.cpp#L1874

Reply:
oha, we still nee to calculate action mroot but just we don't eventually store it in block header.
Will revert back.